Fix crash in dogstatsd

vickenty opened the pull request:
### What does this PR do?

Fix a crash caused by unsynchronized concurrent map and slice accesses in the dogstatsd origin statistics.

### Motivation

Bugfix.

### Additional Notes

With dogstatsd metric stats enabled, when the dogstatsd server receives traffic for the first time, a structure is created to hold telemetry counters for each origin, which is stored in `cachedTlmOriginIds`. If this happens on multiple goroutines at the same time, the agent crashes with the `concurrent map read and map write` error. This condition is made worse when over 200 containers were observed, and old origins are being removed from the map as well.

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

Run the agent with `dogstatsd_metrics_stats_enable: true`.

On docker, start a 5-10 number of containers that start sending metrics at the same time (by timer or another trigger). Agent should not crash.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

vickenty committed:
Fix concurrent map access in dogstatsd stats.

Multiple dogstatsd workers may access and modify cachedTlmOriginIds
and cachedOrder, thus the accesses need to be protected by a lock.

vickenty committed:
Update comments

vickenty committed:
add changelog

vickenty committed:
go fmt

olivielpeau reviewed (commented):
Thanks, can you add details in the PR description about the specific scenarios that can trigger this crash? Both as context for the fix description and to help with a good QA of this change.

Do we have unit tests that could uncover such a race condition? (might be tracked as a separate follow up if it's complex to implement)

remeh committed:
[dogstatsd] add an unit test confirming the absence of data race on `cachedTlmOriginIds`

vickenty committed:
Make the test case more robust

Using distinct origin ids in the test makes createOriginTagMaps
concurrently write to the hash map, triggering a runtime crash, rather
than just a warning with -race.

Sync up all goroutines calls to increase the probability of a fatal
collision.

Waiting for goroutines to finish makes sure that any crash happens
while the test process is still running (if the test is the last in
the queue, the only test running).
